RIOT-OS/RIOT
# gcoap: Add file server

### Contribution description

This is a continuation of @chrysn's GCoAP fileserver implementation with only minimal changes.
Files are served block-wise, directory listings are now too.

My use case is to provide a CoAP fileserver to serve SUIT update files from SD card. 

### Testing procedure

Instead of extending the already cluttered `examples/filesystem`, I added a new `examples/gcoap_fileserver` that leverages `vfs_default` and is thus much cleaner.

To get some data on the (virtual) file system you can use #17937 to download some files from `aiocoap-fileserver`, then if you are on `native` copy `MEMORY.bin` to `examples/gcoap_fileserver`.

You should now be able to request the files via aiocoap-client.

```
main(): This is RIOT! (Version: 2022.07-devel-107-ge39b5-coapfileserver)
> ls /nvm0
ls /nvm0
./
../
LICENSE	24310 B
hello.txt	15 B
total 2 files
```

```
benpicco@T430s ~/dev/RIOT (git)-[coapfileserver] % aiocoap-client coap://[fe80::748f:e6ff:fed7:426d%tapbr0]/vfs/
</vfs/LICENSE>,</vfs/hello.txt>
benpicco@T430s ~/dev/RIOT (git)-[coapfileserver] % aiocoap-client coap://[fe80::748f:e6ff:fed7:426d%tapbr0]/vfs/hello.txt
Hello from RIOT
benpicco@T430s ~/dev/RIOT (git)-[coapfileserver] % aiocoap-client coap://[fe80::748f:e6ff:fed7:426d%tapbr0]/vfs/LICENSE > /tmp/LICENSE
benpicco@T430s ~/dev/RIOT (git)-[coapfileserver] % md5sum LICENSE /tmp/LICENSE 
a4ce26a8dcc018faa99a6cd08cbcd1c3  LICENSE
a4ce26a8dcc018faa99a6cd08cbcd1c3  /tmp/LICENSE
```


### Issues/PRs references

taken over from #14397
